src/android/com/microsoftopentechnologies/intellij/components/MSOpenTechToolsProject.java

```diff
@@ -0,0 +1,172 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.microsoftopentechnologies.intellij.components;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.UnmodifiableIterator;
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Files;
+import com.intellij.openapi.components.AbstractProjectComponent;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.*;
+import com.microsoftopentechnologies.intellij.helpers.ServiceCodeReferenceHelper;
+import com.microsoftopentechnologies.intellij.helpers.UIHelper;
+
+import java.io.*;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.List;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+public class MSOpenTechToolsProject extends AbstractProjectComponent {
+    private static final String TEMPLATES_ROOT_DIR = "build/intermediates/exploded-aar/Microsoft/templates/%s";
+    private static final String TEMPLATES_RESOURCE_PATH = "/com/microsoftopentechnologies/intellij/templates/MobileServiceTemplate/";
+    private static final String TEMPLATE_ZIP_NAME = "templates.zip";
+    private static final String CACHED_TEMPLATE_ZIP_NAME = "templates-%s.zip";
+
+    protected MSOpenTechToolsProject(Project project) {
+        super(project);
+    }
+
+    private VirtualFileListener virtualFileListener = null;
+    private boolean templatesCreated = false;
+
+    @Override
+    public void projectOpened() {
+        try {
+            templatesCreated = false;
+
+            // if a new project is being created then this method is called before
+            // any file is created; we install a file listener to handle the template
+            // file creation in case the project being created is an Android gradle
+            // project
+            virtualFileListener = new VirtualFileAdapter() {
+                @Override
+                public void contentsChanged(VirtualFileEvent event) {
+                    super.fileCreated(event);
+
+                    // if this is a build.gradle file and is an android build file then
+                    // copy the template files
+                    try {
+                        if(!templatesCreated &&
+                                event.getFileName().contains("build.gradle") &&
+                                ServiceCodeReferenceHelper.isAndroidGradleBuildFile(event.getFile())) {
+                            createActivityTemplates();
+                        }
+                    } catch (IOException ignored) {}
+                }
+            };
+            VirtualFileManager.getInstance().addVirtualFileListener(virtualFileListener);
```

Comment:
The VirtualFileManager instance is shared across all open projects. If we have more than one project open at a time, it will run several listeners (in the past, we had a synchronized variable to count the number of references, but register just one listener.) 

If you use only one listener, you can get the  actual project being modified from the event:
VirtualFile vf = event.getFile();
ProjectLocator.getInstance().getProjectsForFile(vf);

With this approach, this logic can be moved to an application component.

You could even create a Map<String,Boolen> with (project name+location , templatesCreated) to handle subsequent modifications to build.gradle.